File: QueueUsingTwoStacks/CPP/QueueUsingTwoStacks.cpp
#include <iostream>
#include <string>

using namespace std;

class Stack
{
private:
    class Node
    {
    public:
        int Value;
        Node *Next;
        Node *Previous;

        Node(){};
        Node(int value)
        {
            Value = value;
        }
    };

    Node *Current;

public:
    int Count;

    Stack()
    {
        Count = 0;
        Current = nullptr;
    }

    void Push(int value)
    {
        if (Count == 0)
        {
            Current = new Node(value);
        }
        else
        {
            Node *Temp = Current;
            Current = new Node(value);
            Current->Previous = Temp;
            Temp->Next = Current;
        }

        Count++;
    }

    int Pop()
    {
        if (Count == 0)
        {
            return -1;
        }

        Count--;
        int value = Current->Value;

        if (!Count)
        {
            delete Current;
        }
        else
        {
            Current = Current->Previous;
            delete Current->Next;
        }

        return value;
    }

    int Peek()
    {
        if (Count == 0)
        {
            return -1;
        }

        return Current->Value;
    }
};

int main()
{
    int cases;
    cin >> cases;

    Stack pushStack;
    Stack popStack;

    for (int i = 0; i < cases; i++)
    {
        string c;
        getline(cin >> ws, c);

        int f = stoi(c.substr(0, c.find(" ")));

        switch (f)
        {
        case 1:
        {
            int num = stoi(c.substr(c.find(" ") + 1, c.size() - 1));
            pushStack.Push(num);
        }
        break;
        case 2:
            if (!popStack.Count)
            {
                while (pushStack.Count)
                {
                    popStack.Push(pushStack.Pop());
                }
            }
            popStack.Pop();
            break;
        case 3:
            if (!popStack.Count)
            {
                while (pushStack.Count)
                {
                    popStack.Push(pushStack.Pop());
                }
            }
            cout << popStack.Peek() << "\n";
            break;
        }
    }

    return 0;
}

File: QueueUsingTwoStacks/Python/QueueUsingTwoStacks.py
#!/bin/python3


class Stack:
    class __Node:
        value = None
        next = None
        previous = None

        def __init__(self, value):
            self.value = value

    __current = None
    count = None

    def __init__(self):
        self.count = 0

    def push(self, value):
        if self.count == 0:
            self.__current = self.__Node(value)
        else:
            temp = self.__current
            self.__current = self.__Node(value)
            self.__current.previous = temp
            temp.next = self.__current

        self.count += 1

    def pop(self):
        if self.count == 0:
            return -1

        self.count -= 1
        value = self.__current.value

        if self.count == 0:
            self.__current = None
        else:
            self.__current = self.__current.previous
            self.__current.next = None

        return value

    def peek(self):
        if self.count == 0:
            return -1

        return self.__current.value


if __name__ == '__main__':
    cases = int(input())

    pushStack = Stack()
    popStack = Stack()

    for case in range(cases):
        c = input()
        split = c.split(' ')
        f = int(split[0])

        if f == 1:
            pushStack.push(int(split[1]))
        elif f == 2:
            if popStack.count == 0:
                while pushStack.count > 0:
                    popStack.push(pushStack.pop())
            popStack.pop()
        elif f == 3:
            if popStack.count == 0:
                while pushStack.count > 0:
                    popStack.push(pushStack.pop())
            print(popStack.peek())
